### PR TITLE
Tech-debt: Select\Multiselect\Filter fields refactor

### LINKED ISSUE #503: [Refactor] Break up the `Select` control component into multiple components
**What needs to be done? Why does it need to be done?**
Break the `src/components/controls/Select` component into basic `Select` and `MultiSelect` components. That component is quite complex and breaking them out will reduce the amount of and complexity of logic significantly.

I would also suggest pulling `ValueContainer`, `GroupHeading`, and `Option` into their own component files as well, which will pull their helper functions with them, and `MultiSelect` will end up being nice and clean.

**Additional context**
Add any other context about the task here.

### PR DESCRIPTION
**Pull request by Wmaileq**
## Description of the change
**Need to be merged after #541** 
- Select separated into MultiSelect and Select
- Filters that use Select refactored and generalized
- Filter types (reportedViolations, violationType, etc.) picked out into constants. 

## Type of change

- [x] Bug fix (non-breaking change that fixes an issue)
- [ ] New feature (non-breaking change that adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Configuration change (adjusts configuration to achieve some end related to functionality, development, performance, or security)

## Related issues

> Closes #503 

### Development

These boxes should be checked by the submitter prior to merging:

- [x] Manual testing against realistic data has been performed locally

### Code review

These boxes should be checked by reviewers prior to merging:

- [ ] This pull request has a descriptive title and information useful to a reviewer
- [ ] This pull request has been moved out of a Draft state, has no "Work In Progress" label, and has assigned reviewers
- [ ] Potential security implications or infrastructural changes have been considered, if relevant

### CODE REVIEW

**Comment by jovergaag** [comment]
@Wmaileq I am partially through reviewing this, I'll pick it up in the morning when I have a fresh brain.

File: src/components/charts/new_revocations/ToggleBar/AdmissionTypeFilter.js [hunk 1]
```diff
@@ -15,29 +15,37 @@
...
+import { flatOptions } from "../../../controls/utils";
 
 const AdmissionTypeFilter = ({
   defaultValue,
+  value,
   summingOption,
   onChange,
   options = [],
 }) => {
+  const onValueChange = useCallback(
```
Comment (jovergaag) [inline src/components/charts/new_revocations/ToggleBar/AdmissionTypeFilter.js:34]: Great idea moving toward optimizing! I only looked into this briefly but it looks like the memoization of the child is still being broken. You may need to use React.memo on those children if we really want to take advantage of these memo hooks.

Comment (jovergaag) [reply src/components/charts/new_revocations/ToggleBar/AdmissionTypeFilter.js:34]: Actually maybe removing `onChange` from the deps would fix it? Not sure, something to look into.

File: src/components/toggles/ToggleBar.js [hunk 2]
```diff
@@ -17,12 +17,13 @@
 
 /* eslint-disable jsx-a11y/label-has-associated-control */
 
-import React, { useMemo } from "react";
+import React, { useCallback, useMemo } from "react";
```
Comment (jovergaag) [inline src/components/toggles/ToggleBar.js:20]: It doesn't need to happen in this PR but let's make a mental note to somehow get rid of this component and merge it with the other `ToggleBar`. It's super confusing to have both of these, especially for someone new to the repo!

File: src/components/controls/utils.js [hunk 3]
```diff
@@ -0,0 +1,118 @@
...
+ [
+ { value: "All", label: "ALL" },
+ { value: "REVOCATION", label: "Revocation" },
+ { value: "SCI_6", label: "SCI 6 months" },
+ { value: "DA_DETOX", label: "D&A Detox" },
+ ]
+ * @param options
+ * @returns {*}
+ */
+export const flatOptions = (options) =>
```
Comment (jovergaag) [inline src/components/controls/utils.js:31]: Beautiful

File: src/components/charts/new_revocations/ToggleBar/DistrictFilter.js [hunk 4]
```diff
@@ -50,72 +51,79 @@ const DistrictFilter = ({ stateCode, onChange }) => {
...
-  if (district) {
-    return (
-      <FilterField label="District">
+  const onValueChange = useCallback(
+    (options) => onChange(map("value", options)),
+    [onChange]
+  );
+
+  const select = useMemo(() => {
+    if (district) {
```
Comment (jovergaag) [inline src/components/charts/new_revocations/ToggleBar/DistrictFilter.js:60]: Have you tested this user context branch to verify it is working as expected? We removed the region condition below so just want to make sure you checked that.

**Review by jovergaag — CHANGES_REQUESTED** [review]
Great work @Wmaileq! Overall I think it looks great. It feels much more clean to me broken out into multiple components. I also really like the move toward constants and React hooks for optimization.

Just a few things in there to look at. 

Another thing that I noticed. The `MultiSelect` functionality seems to have regressed in IE11. It now no longer has the multi select function, and closes the drop down after a single select. 

Thanks!

File: src/components/charts/new_revocations/ToggleBar/DistrictFilter.js (hunk 4 above)
Comment (Wmaileq) [reply src/components/charts/new_revocations/ToggleBar/DistrictFilter.js:60]: Just checked it, all works as expected.

File: src/components/charts/new_revocations/ToggleBar/AdmissionTypeFilter.js (hunk 1 above)
Comment (Wmaileq) [reply src/components/charts/new_revocations/ToggleBar/AdmissionTypeFilter.js:34]: As perfomance optimizations is not target of this PR i've reverted it and, maybe, will come back later.

File: src/components/charts/new_revocations/ToggleBar/ToggleBarFilter.js [hunk 5]
```diff
@@ -15,36 +15,44 @@
...
 
-import React from "react";
+import React, { useCallback } from "react";
 import PropTypes from "prop-types";
 
 import Select from "../../../controls/Select";
 import FilterField from "./FilterField";
+import { optionPropType } from "../../../propTypes";
+
+const ToggleBarFilter = ({
```
[the remark this answers is not in the document]
Comment (Wmaileq) [reply src/components/charts/new_revocations/ToggleBar/ToggleBarFilter.js:25]: As long as i remember this was done in another PR: https://github.com/Recidiviz/pulse-dashboard/pull/512/files#diff-4246339bd834789c3b3212cee11462ceL33

**Review by jovergaag — CHANGES_REQUESTED** [review]
Couple more little styling fixes.

File: src/components/charts/new_revocations/ToggleBar/AdmissionTypeFilter.js [hunk 6]
```diff
@@ -31,10 +31,7 @@ const AdmissionTypeFilter = ({
   onChange,
   options = [],
 }) => {
-  const onValueChange = useCallback(
-    (selected) => onChange(map("value", selected)),
-    [onChange]
-  );
+  const onValueChange = (selected) => onChange(map("value", selected));
```
[the remark this answers is not in the document]
Comment (Wmaileq) [reply src/components/charts/new_revocations/ToggleBar/AdmissionTypeFilter.js:34]: No issues, this hooks work as expected but do bring a little benefit while corresponding components and linked vars and callbacks are not wrapped with memo's, useCallback's and useMemo's
I think we can do this optimization in another PR and test if perfomance becomes faster or not.

File: src/components/controls/MultiSelect/MultiSelect.scss [hunk 7]
```diff
@@ -1,6 +1,10 @@
 .MultiSelect {
   flex-grow: 1;
 
+  &__value-container {
```
[the remark this answers is not in the document]
Comment (Wmaileq) [reply src/components/controls/MultiSelect/MultiSelect.scss:4]: > IE11 still seems to be broken for the Lantern dashboards. ND looks ok.

Yes, did not have windows PC at hand, testing IE11 right now :)

[the remark this answers is not in the document]
Comment (Wmaileq) [reply src/components/controls/MultiSelect/MultiSelect.scss:4]: Fixed IE11 bug(option selecting does not close select) and fixed width of all filter fields (now distributing evenly)

File: src/components/charts/new_revocations/ToggleBar/FilterField.scss [hunk 8]
```diff
@@ -3,6 +3,7 @@
   flex-direction: column;
   padding: 1.5rem 3rem;
   flex-grow: 1;
+  flex-basis: 100%;
```
Comment (jovergaag) [inline src/components/charts/new_revocations/ToggleBar/FilterField.scss:6]: The product team has been through a round of QA and has a suggestion on how to make the district filter look nice on tablet. I will put this in another issue so we can get this PR merged, just letting you know that it is coming.

File: src/components/controls/MultiSelect/MultiSelect.js [hunk 9]
```diff
@@ -41,9 +41,6 @@ const MultiSelect = ({
     (selectedOptions) => {
       const newOptions = getNewOptions(options, summingOption, selectedOptions);
       onChange(newOptions);
-      setTimeout(() => {
```
[the remark this answers is not in the document]
Comment (jovergaag) [reply src/components/controls/MultiSelect/MultiSelect.js]: Regardless of whether the above helps or not (I did not try it on your branch), we need to make sure the input gets focused after each selection, and that the dropdown does not close after each selection until the user closes it.

**Review by jovergaag — CHANGES_REQUESTED** [review]

File: src/components/controls/MultiSelect/MultiSelect.js (hunk 9 above)
[the remark this answers is not in the document]
Comment (Wmaileq) [reply src/components/controls/MultiSelect/MultiSelect.js]: I've reverted this and tested in IE11 emulator, safari and chrome and multiselect works as expected.

**Comment by Wmaileq** [comment]
As #541 and this intersect each other, we should wait till #541 would be merged, then merge master here and then we can close PR

**Review by jovergaag — APPROVED** [review]
Looks great! Thank you.

**Pull request merged**